Q1. A company conducted a survey of its employees and found that 70% of the employees use the
company's health insurance plan, while 40% of the employees who use the plan are smokers. What is the
probability that an employee is a smoker given that he/she uses the health insurance plan?

Ans.

P(B|A) = P(A ∩ B) / P(A) = 0.28 / 0.7 = 0.4
The probability that an employee is a smoker given that they use the health insurance plan is 40%.

Q2. What is the difference between Bernoulli Naive Bayes and Multinomial Naive Bayes?

Ans.
The main difference between Bernoulli Naive Bayes and Multinomial Naive Bayes is the way they handle categorical features. Bernoulli Naive Bayes treats each categorical feature as a binary variable, where 1 indicates the presence of the feature and 0 indicates its absence. Multinomial Naive Bayes, on the other hand, treats each categorical feature as a multinomial variable, where the value of the feature can be any one of the possible categories.

Q3. How does Bernoulli Naive Bayes handle missing values?

Ans.
Bernoulli Naive Bayes handles missing values by simply dropping the feature from the model. This is because Bernoulli Naive Bayes assumes that each feature is binary, and a missing value cannot be represented as a binary value.

Q4. Can Gaussian Naive Bayes be used for multi-class classification?

Ans.
Yes, Gaussian Naive Bayes can be used for multi-class classification. This is because Gaussian Naive Bayes assumes that the features are normally distributed, and the distribution of each feature can be different for each class.


In [ ]:
Q5. Assignment:
    
Data preparation:
Download the "Spambase Data Set" from the UCI Machine Learning Repository (https://archive.ics.uci.edu/ml/
datasets/Spambase). This dataset contains email messages, where the goal is to predict whether a message
is spam or not based on several input features.

Implementation:
Implement Bernoulli Naive Bayes, Multinomial Naive Bayes, and Gaussian Naive Bayes classifiers using the
scikit-learn library in Python. Use 10-fold cross-validation to evaluate the performance of each classifier on the
dataset. You should use the default hyperparameters for each classifier.

Results:
Report the following performance metrics for each classifier:
Accuracy
Precision
Recall
F1 score

Discussion:
Discuss the results you obtained. Which variant of Naive Bayes performed the best? Why do you think that is
the case? Are there any limitations of Naive Bayes that you observed?

Conclusion:
Summarise your findings and provide some suggestions for future work.

In [3]:
import pandas as pd
url = "https://archive.ics.uci.edu/ml/machine-learning-databases/spambase/spambase.data"
df = pd.read_csv(url, header=None)

In [4]:
df.head()

0     1     2    3     4     5     6     7     8     9   ...    48  \
0  0.00  0.64  0.64  0.0  0.32  0.00  0.00  0.00  0.00  0.00  ...  0.00   
1  0.21  0.28  0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
2  0.06  0.00  0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
3  0.00  0.00  0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00  0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   

      49   50     51     52     53     54   55    56  57  
0  0.000  0.0  0.778  0.000  0.000  3.756   61   278   1  
1  0.132  0.0  0.372  0.180  0.048  5.114  101  1028   1  
2  0.143  0.0  0.276  0.184  0.010  9.821  485  2259   1  
3  0.137  0.0  0.137  0.000  0.000  3.537   40   191   1  
4  0.135  0.0  0.135  0.000  0.000  3.537   40   191   1  

[5 rows x 58 columns]

In [9]:
X = df.iloc[:,:-1]
y = df.iloc[:,-1]

In [10]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.20 , random_state = 42)

In [6]:
from sklearn.naive_bayes import BernoulliNB, MultinomialNB, GaussianNB
bn = BernoulliNB()
mn = MultinomialNB()
gb = GaussianNB()

In [13]:
gb.fit(X_train,y_train)

GaussianNB()

In [15]:
y_pred = gb.predict(X_test)

In [16]:
y_pred

array([1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1,
       1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0,
       0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1, 1,
       1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1,
       1, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 1, 1,
       0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0,
       1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1,
       1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0,
       0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0,

In [17]:
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix
print(accuracy_score(y_test,y_pred))
print(classification_report(y_test,y_pred))
print(confusion_matrix(y_test,y_pred))

0.8208469055374593
              precision    recall  f1-score   support

           0       0.95      0.73      0.82       531
           1       0.72      0.95      0.82       390

    accuracy                           0.82       921
   macro avg       0.83      0.84      0.82       921
weighted avg       0.85      0.82      0.82       921

[[387 144]
 [ 21 369]]


In [21]:
mn.fit(X_train,y_train)

MultinomialNB()

In [24]:
y_pred1 = mn.predict(X_test)

In [25]:
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix
print(accuracy_score(y_test,y_pred1))
print(classification_report(y_test,y_pred1))
print(confusion_matrix(y_test,y_pred1))

0.7861020629750272
              precision    recall  f1-score   support

           0       0.80      0.84      0.82       531
           1       0.76      0.72      0.74       390

    accuracy                           0.79       921
   macro avg       0.78      0.78      0.78       921
weighted avg       0.79      0.79      0.79       921

[[445  86]
 [111 279]]


In [28]:
bn.fit(X_train,y_train)

BernoulliNB()

In [29]:
y_pred2 = bn.predict(X_test)

In [30]:
from sklearn.metrics import accuracy_score,classification_report,confusion_matrix
print(accuracy_score(y_test,y_pred2))
print(classification_report(y_test,y_pred2))
print(confusion_matrix(y_test,y_pred2))

0.8805646036916395
              precision    recall  f1-score   support

           0       0.86      0.94      0.90       531
           1       0.91      0.80      0.85       390

    accuracy                           0.88       921
   macro avg       0.89      0.87      0.88       921
weighted avg       0.88      0.88      0.88       921

[[499  32]
 [ 78 312]]


In [31]:
from sklearn.model_selection import KFold

In [32]:
cv = KFold(n_splits = 10)

In [33]:
cv

KFold(n_splits=10, random_state=None, shuffle=False)

In [41]:
import numpy as np
from sklearn.model_selection import cross_val_score
score_bn = cross_val_score(bn,X_train,y_train,cv = cv , scoring = 'accuracy')
print(score_bn)

print(np.mean(score_bn))


[0.87228261 0.91032609 0.86413043 0.88043478 0.89402174 0.85597826
 0.88586957 0.89673913 0.88043478 0.91847826]
0.8858695652173912


In [43]:
score_mn = cross_val_score(mn,X_train,y_train,cv = cv , scoring = 'accuracy')
print(score_mn)
print(np.mean(score_mn))


[0.77173913 0.80434783 0.76630435 0.79347826 0.77173913 0.79891304
 0.80706522 0.82608696 0.80978261 0.77445652]
0.7923913043478261


In [42]:
score_gb = cross_val_score(gb,X_train,y_train,cv = cv , scoring = 'accuracy')
print(score_gb)
print(np.mean(score_gb))


[0.82880435 0.83152174 0.76630435 0.81793478 0.82336957 0.82065217
 0.81521739 0.83967391 0.81793478 0.83695652]
0.819836956521739


Conclusion - Which Variant Performed the Best:
    Bernoulli naive bayes performed the best with an accuracy of 88%


Reason for Bernoulli Naive Bayes Performance:
The success of Bernoulli Naive Bayes could be attributed to the nature of the dataset. Since the dataset involves email messages, the binary representation of features might capture the presence or absence of specific words or patterns that are indicative of spam. Bernoulli Naive Bayes excels when dealing with binary features like these, making it effective for text classification tasks such as spam detection.

Limitations of Naive Bayes:
While Bernoulli Naive Bayes showed strong performance in this scenario, it's important to acknowledge the limitations of Naive Bayes algorithms:

Feature Independence Assumption: Naive Bayes assumes that features are independent given the class label. In real-world datasets, this assumption might not hold true, leading to potential inaccuracies.

Sensitivity to Irrelevant Features: Naive Bayes treats all features as equally important. If there are irrelevant or noisy features, they might negatively impact the performance.

Zero Probability Issue: When a feature-class combination is not observed in the training data, Naive Bayes assigns a zero probability, causing the entire likelihood to become zero and predictions to be unreliable.

Conclusion and Future Work:
In conclusion, Bernoulli Naive Bayes proved to be the most suitable variant for the "Spambase" dataset, achieving an accuracy of 88%. However, further exploration is needed to improve performance and overcome the limitations of Naive Bayes.

Future work could include:

Feature Engineering: Experiment with different feature representations, including techniques that consider word frequencies or other relevant features.

Data Preprocessing: Address missing values, remove noise, and preprocess the text data effectively to enhance classification accuracy.

Hyperparameter Tuning: Fine-tune hyperparameters to potentially enhance the performance of the Naive Bayes classifiers.

Ensemble Techniques: Explore ensemble methods to combine predictions from multiple classifiers, potentially boosting overall accuracy.
